packages/rln/src/contract/rln_contract.spec.ts

```diff
@@ -1,82 +1,277 @@
-      rlnInstance.zerokit.generateSeededIdentityCredential(mockSignature);
-    await rlnContract.registerWithIdentity(identity);
+    const mockedRegistryContract = {
+      register: registerStub,
+      queryFilter: () => [membershipRegisteredEvent],
+      provider: {
+        getLogs: () => [],
+        getBlockNumber: () => Promise.resolve(1000),
+        getNetwork: () => Promise.resolve({ chainId: 11155111 })
+      },
+      address: SEPOLIA_CONTRACT.address,
+      interface: {
+        getEvent: (eventName: string) => ({
+          name: eventName,
+          format: () => {}
+        })
+      },
+      filters: {
+        MembershipRegistered: () => ({}),
+        MembershipRemoved: () => ({})
+      },
+      on: () => ({}),
+      removeAllListeners: () => ({})
+    };
+
+    const provider = new ethers.providers.JsonRpcProvider();
+    const voidSigner = new ethers.VoidSigner(
+      SEPOLIA_CONTRACT.address,
+      provider
+    );
+    const rlnContract = await RLNContract.init(rlnInstance, {
+      signer: voidSigner,
+      address: SEPOLIA_CONTRACT.address,
+      rateLimit: DEFAULT_RATE_LIMIT,
+      contract: mockedRegistryContract as unknown as ethers.Contract
+    });
+
+    const decryptedCredentials =
+      await rlnContract.registerWithIdentity(identity);
+
+    expect(decryptedCredentials).to.not.be.undefined;
+    if (!decryptedCredentials)
+      throw new Error("Decrypted credentials should not be undefined");
 
-    expect(registerCalled).to.be.true;
+    expect(
+      registerStub.calledWith(
+        identity.IDCommitmentBigInt,
+        DEFAULT_RATE_LIMIT,
+        [],
+        {
+          gasLimit: 300000
+        }
+      )
+    ).to.be.true;
+
+    expect(decryptedCredentials).to.have.property("identity");
+    expect(decryptedCredentials).to.have.property("membership");
+    expect(decryptedCredentials.membership).to.include({
+      address: SEPOLIA_CONTRACT.address,
+      treeIndex: 1
+    });
+
+    const expectedIdCommitment = ethers.utils.zeroPad(
+      hexToBytes(formatIdCommitment(identity.IDCommitmentBigInt)),
+      32
+    );
+    expect(insertMemberSpy.callCount).to.equal(2);
+    expect(insertMemberSpy.getCall(1).args[0]).to.deep.equal(
+      expectedIdCommitment
+    );
+  });
+
+  describe("Rate Limit Management", () => {
+    it("should get contract rate limit parameters", async () => {
+      const minRate = await rlnContract.getMinRateLimit();
+      const maxRate = await rlnContract.getMaxRateLimit();
+      const maxTotal = await rlnContract.getMaxTotalRateLimit();
+      const currentTotal = await rlnContract.getCurrentTotalRateLimit();
+
+      expect(minRate).to.equal(mockRateLimits.minRate);
+      expect(maxRate).to.equal(mockRateLimits.maxRate);
+      expect(maxTotal).to.equal(mockRateLimits.maxTotalRate);
+      expect(currentTotal).to.equal(mockRateLimits.currentTotalRate);
+    });
+
+    it("should calculate remaining total rate limit", async () => {
```

Comment:
what is this test about? 
it looks that we are checking for leftover rate limit BUT there were no operations being done in that suit

Reply:
You're right about this test. It's correctly verifying that `getRemainingTotalRateLimit()` calculates `maxTotalRate - currentTotalRate`, but it's only testing the calculation logic with pre-mocked values (1000 - 500 = 500).
The test is technically correct but limited in scope - it doesn't test how rate limits change after operations. We could improve it by adding tests that simulate registrations/removals and verify the rate limit updates accordingly.
For now, this test ensures our calculation method works as expected with the mocked contract values.